### PR TITLE
Upgrades: Golang 1.19, Knative v1.9, K8s v1.25.3, Istio 1.16.0, metallb 0.3.19

### PR DESCRIPTION
**Pull request by anshalshukla**
List of upgrades:
* Golang to 1.19
* Knative to 1.9
* k8s to 1.25
* Istio to 1.16
* metallb to 0.3.19

WARNING: After this PR, vHive is incompatible with k8s <1.24

### CODE REVIEW

**Comment by leokondrashov** [comment]
I've ran the PR branch through the vHive quickstart guide. Looks fine, in the end invoker runs as expected.

Should there be any other checks before merge?

File: scripts/cluster/setup_master_node.sh [hunk 1]
```diff
@@ -48,7 +48,7 @@ export PATH=$PATH:$ROOT/istio-1.12.5/bin
 sudo sh -c  "echo 'export PATH=\$PATH:$ROOT/istio-1.12.5/bin' >> /etc/profile"
 istioctl install -y -f $ROOT/configs/istio/istio-minimal-operator.yaml
```
Comment (cvetkovic) [inline scripts/cluster/setup_master_node.sh:49]: Please do update Istio as well if you already updated K8s and Knative. 
You need to find the proper version of K8s, Knative, and Istio that all support each other. 

Here is the compatibility table:
- https://istio.io/latest/docs/releases/supported-releases/

It also looks like the deployment method for Istio has changed, hence you should adapt the script:
- https://github.com/knative-sandbox/net-istio/releases
- https://knative.dev/docs/install/installing-istio/#basic-installation-with-istioctl

File: docs/developers_guide.md [hunk 2]
```diff
@@ -185,7 +185,7 @@ Here are some useful commands (there are plenty of Zipkin tutorials online):
 of our [fork](https://github.com/vhive-serverless/firecracker-containerd) of the upstream repository.
 Currently, we are in the process of upstreaming VM snapshots support to the upstream repository.
 
-* Current Firecracker version is 0.24.0, Knative 1.4, Kubernetes 1.23.5, gVisor 20210622.0, and Istio 1.12.5.
+* Current Firecracker version is 0.24.0, Knative 1.9, Kubernetes 1.24.0, gVisor 20210622.0, and Istio 1.12.5.
```
Comment (cvetkovic) [inline docs/developers_guide.md:188]: Update Istio.

File: CHANGELOG.md [hunk 3]
```diff
@@ -14,7 +14,7 @@
 ### Added
 
 ### Changed
-- Bumped Containerd to v1.6, Knative to v1.4, Go to v1.18, Kubernetes to v1.23, Istio to 1.12, protoc to 3.19, runc to 1.1.
+- Bumped Containerd to v1.6, Knative to v1.9, Go to v1.18, Kubernetes to v1.24, Istio to 1.12, protoc to 3.19, runc to 1.1.
```
Comment (cvetkovic) [inline CHANGELOG.md:17]: Update Istio.

**Review by cvetkovic — CHANGES_REQUESTED** [review]

File: scripts/cluster/setup_master_node.sh [hunk 4]
```diff
@@ -48,7 +48,7 @@ export PATH=$PATH:$ROOT/istio-1.12.5/bin
...
 
-KNATIVE_VERSION="knative-v1.4.0"
+KNATIVE_VERSION="knative-v1.9.0"
```
Comment (cvetkovic) [inline scripts/cluster/setup_master_node.sh:51]: Did you modify the queue-proxy of Knative v1.9 prior to building it? If not, Firecracker mode won't work.

You need to sync the fork of Knative (https://github.com/ease-lab/serving) and checkout to v1.9 and apply the following changes:
- https://github.com/ease-lab/serving/commit/8ab7e6985be0bcf8263c347a6f02148e492f63be
- https://github.com/ease-lab/serving/commit/4aac0b2b5b1bb901d69297083d2e6421ad659e02

Then you can build Knative and change the YAML files in `configs/knative_yamls`.

**Review by cvetkovic — CHANGES_REQUESTED** [review]

File: scripts/cluster/setup_master_node.sh (hunk 4 above)
Comment (anshalshukla) [reply scripts/cluster/setup_master_node.sh:52]: Yes, I did modify the queue-proxy of Knative v1.9, I've updated the Knative upgrade [Wiki ](https://github.com/vhive-serverless/vHive/wiki/TMP:-How-to-Update-Knative-for-vHive) as well with the appropriate instructions. I've generate the yamls with the changes on this [PR](https://github.com/ease-lab/serving/pull/10).

File: scripts/github_runner/Dockerfile.cri_dev_env [hunk 5]
```diff
@@ -20,7 +20,7 @@
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 # SOFTWARE.
 
-FROM kindest/node:v1.23.5
+FROM kindest/node:v1.24.0
```
Comment (cvetkovic) [inline scripts/github_runner/Dockerfile.cri_dev_env:23]: Should this be `v1.25.6`?

File: scripts/cluster/setup_master_node.sh (hunk 4 above)
Comment (cvetkovic) [reply scripts/cluster/setup_master_node.sh:52]: Looks good. Have you updated YAMLs in this PR to include changes in the queue-proxy? Once that is done, please test the whole setup with instructions from the quick start guide, insect statuses of pods in K8s, check there aren't errors (post the output here), and run some workload. If that passes, we are good to merge the PR.

**Review by cvetkovic — CHANGES_REQUESTED** [review]

File: scripts/setup_zipkin.sh [hunk 6]
```diff
@@ -24,7 +24,7 @@
 
 set -e
 # install zipkin pods
-kubectl apply -f https://raw.githubusercontent.com/istio/istio/release-1.12/samples/addons/extras/zipkin.yaml
+kubectl apply -f https://raw.githubusercontent.com/istio/istio/release-1.16.0/samples/addons/extras/zipkin.yaml
```
Comment (cvetkovic) [inline scripts/setup_zipkin.sh:27]: This file does not exist.
https://github.com/vhive-serverless/vHive/actions/runs/4222257782/jobs/7330544761#step:9:481

**Comment by ustiugov** [comment]
@anshalshukla please investigate the failing unit test (it probably also causes the CRI tests to fail). Also, please squash your commits that belong to the same logical changes into one using git rebase

File: scripts/cluster/setup_master_node.sh (hunk 4 above)
Comment (anshalshukla) [reply scripts/cluster/setup_master_node.sh:52]: Yes, I did update the YAMLs in this PR. I've tested it and everything works as expected.

**Review by cvetkovic — DISMISSED** [review]
LGTM!

File: configs/kind/kubeadm.conf [hunk 7]
```diff
@@ -1,12 +1,12 @@
...
 kind: ClusterConfiguration
-kubernetesVersion: v1.23.5
+kubernetesVersion: v1.25.3
 networking:
   podSubnet: 192.168.0.0/16
 scheduler:
   extraArgs: null
 ---
-apiVersion: kubelet.config.k8s.io/v1beta1
+apiVersion: kubelet.config.k8s.io/v1beta3
```
Comment (ustiugov) [inline configs/kind/kubeadm.conf:9]: Shouldn't it be `v1`?

File: CHANGELOG.md [hunk 8]
```diff
@@ -26,7 +26,7 @@
 ### Added
 
 ### Changed
-- Bumped Containerd to v1.6, Knative to v1.4, Go to v1.18, Kubernetes to v1.23, Istio to 1.12, protoc to 3.19, runc to 1.1.
+- Bumped Containerd to v1.6, Knative to v1.9, Go to v1.18, Kubernetes to v1.25.3, Istio to 1.16.0, protoc to 3.19, runc to 1.1.
```
Comment (ustiugov) [inline CHANGELOG.md:29]: add metallb as well

File: cri/cri_test.go [hunk 9]
```diff
@@ -56,20 +56,20 @@ func TestSingleInvoke(t *testing.T) {
 	invoke(t, functionURL)
 }
 
-func TestSingleInvokeLocal(t *testing.T) {
-	functionURL := getFuncURL("helloworldlocal")
-	invoke(t, functionURL)
-}
+// func TestSingleInvokeLocal(t *testing.T) {
```
Comment (ustiugov) [inline cri/cri_test.go:59]: add an entry to the changelog saying that we excluded local registry tests for now, raise an issue to keep track and add the issue number next to the commented code

File: scripts/setup_cluster_and_test_cri.sh [hunk 10]
```diff
@@ -0,0 +1,50 @@
+#!/bin/bash
```
Comment (ustiugov) [inline scripts/setup_cluster_and_test_cri.sh:1]: this must be in a different PR

**Review by ustiugov — CHANGES_REQUESTED** [review]

File: configs/kind/kubeadm.conf (hunk 7 above)
Comment (anshalshukla) [reply configs/kind/kubeadm.conf:9]: I'll rather drop the related commit, these config were related to fixing CI runner within the containers

**Review by ustiugov — CHANGES_REQUESTED** [review]
- [ ] revert the RNN function stuff
- [ ] Clean up the commit history
- [ ] Host knative images under ghcr.io of the vhive org

File: function-images/tests/save_load_minio/server/go.mod [hunk 11]
```diff
@@ -1,13 +1,33 @@
...
 
-go 1.16
+go 1.19
 
 replace github.com/vhive-serverless/vhive/function-images/tests/save_load_minio/proto => ../proto_gen
 
 require (
-	github.com/containerd/containerd v1.4.3
-	github.com/vhive-serverless/vhive/function-images/tests/save_load_minio/proto v0.0.0-00010101000000-000000000000
+	github.com/containerd/containerd v1.6.20
```
Comment (anshalshukla) [inline function-images/tests/save_load_minio/server/go.mod:8]: Can we upgrade this to 1.6.20 without the firecracker-containerd upgrade?
